support-workers/src/main/scala/com/gu/support/workers/lambdas/CreateZuoraSubscription.scala

```diff
@@ -111,61 +116,64 @@ class CreateZuoraSubscription(servicesProvider: ServiceProvider = ServiceProvide
       state.acquisitionData
     )
 
-  private def buildSubscribeItem(state: CreateZuoraSubscriptionState, promotionService: PromotionService): Either[PromoError, SubscribeItem] = {
+  private def buildSubscribeItem(state: CreateZuoraSubscriptionState, subscriptionData: SubscriptionData): SubscribeItem = {
     val billingEnabled = state.paymentMethod.isLeft
-    buildSubscriptionData(state, promotionService).map { subscriptionData =>
-      //Documentation for this request is here: https://www.zuora.com/developer/api-reference/#operation/Action_POSTsubscribe
-      SubscribeItem(
-        account = buildAccount(state),
-        billToContact = buildContactDetails(state.user, None, state.user.billingAddress),
-        soldToContact = state.user.deliveryAddress map (buildContactDetails(state.user, state.giftRecipient, _, state.user.deliveryInstructions)),
-        paymentMethod = state.paymentMethod.left.toOption,
-        subscriptionData = subscriptionData,
-        subscribeOptions = SubscribeOptions(billingEnabled, billingEnabled)
-      )
-    }
+    //Documentation for this request is here: https://www.zuora.com/developer/api-reference/#operation/Action_POSTsubscribe
+    SubscribeItem(
+      account = buildAccount(state),
+      billToContact = buildContactDetails(state.user, None, state.user.billingAddress),
+      soldToContact = state.user.deliveryAddress map (buildContactDetails(state.user, state.giftRecipient, _, state.user.deliveryInstructions)),
+      paymentMethod = state.paymentMethod.left.toOption,
+      subscriptionData = subscriptionData,
+      subscribeOptions = SubscribeOptions(generateInvoice = billingEnabled, processPayments = billingEnabled)
+    )
   }
 
-  private def buildSubscriptionData(state: CreateZuoraSubscriptionState, promotionService: PromotionService) = {
+  private def buildSubscriptionData(
+    state: CreateZuoraSubscriptionState,
+    promotionService: => PromotionService,
+    now: () => LocalDate
+  ): Future[SubscriptionData] = {
     val isTestUser = state.user.isTestUser
     val config = zuoraConfigProvider.get(isTestUser)
-    val readerType: ReaderType = state.giftRecipient match  {
-      case _: Some[GiftRecipient] => ReaderType.Gift
-      case _ => ReaderType.Direct
-    }
-    val stage = Configuration.stage
+    val environment = TouchPointEnvironments.fromStage(Configuration.stage, isTestUser)
 
-    state.product match {
-      case c: Contribution => Right(c.build(state.requestId, config))
-      case d: DigitalPack => d.build(
+    val eventualErrorOrSubscriptionData = state.product match {
+      case c: Contribution => EitherT.pure[Future, Throwable](buildContributionSubscription(c, state.requestId, config))
```

Comment:
it's slightly premature to make it into EitherT rather than just Either, as we don't use Futures yet, but I thought it would simplify the next PR to remove all that types noise.

Reply:
perhaps it's time for zio :)

Reply:
I'd certainly be happy to consider it, or any other non "wtf" type of abstraction of effects!
![image](https://user-images.githubusercontent.com/7304387/84513551-80ee7c00-acc1-11ea-9503-b03cdf57a4ad.png)
Would be great to have a stronger guarantee of not accidentally pushing data to prod during IT tests, and simpler unit testability.
Not sure how we'd justify it as a quality objective though!